test/stream_transformers/buffer_test.dart

```diff
@@ -0,0 +1,138 @@
+import 'dart:async';
+
+import 'package:test/test.dart';
+
+import 'package:async/stream_transformers.dart';
+
+void main() {
+  test('does not emit before `trigger`', () async {
+    var trigger = new StreamController();
+    var values = new Stream.fromIterable([1, 2, 3]);
+    var hasEmitted = false;
+    values.transform(buffer(trigger.stream)).listen((_) {
+      hasEmitted = true;
+    });
+    await new Future(() {});
+    expect(hasEmitted, false);
+    trigger.add(null);
+    await new Future(() {});
+    expect(hasEmitted, true);
+  });
+
+  test('emits immediately if trigger emits before a value', () async {
+    var trigger = new StreamController();
+    var values = new StreamController();
+    trigger.add(null);
+    var hasEmitted = false;
+    values.stream.transform(buffer(trigger.stream)).listen((_) {
+      hasEmitted = true;
+    });
+    values.add(1);
+    await new Future(() {});
+    expect(hasEmitted, true);
+  });
```

Comment:
Looping around all combinations of broadcast/single subscription and running all of the tests now.